core/common/src/main/java/alluxio/file/PositionReadable.java

```diff
@@ -0,0 +1,102 @@
+
+import java.io.Closeable;
+import java.io.IOException;
+import java.nio.ByteBuffer;
+
+/**
+ * Position read interface. All implementer should be thread-safe.
+ */
+// TODO(lu) change to simple interface and have default implementation in abstract class
+// after block store is removed so that PagedFileReader does not need to extend abstract class
+// BlockReader
+public interface PositionReadable extends Closeable {
+  /**
+   * @param position position of the file to start reading data
+   * @param buffer target byte array
+   * @param length bytes to read
+   * @return bytes read, or -1 none of data is read
+   */
+  default int read(long position, byte[] buffer, int length)
+      throws IOException {
+    return read(position, new ByteArrayTargetBuffer(buffer, 0), length);
+  }
+
+  /**
+   * @param position position of the file to start reading data
+   * @param buffer target byte array
+   * @param offset the offset of the buffer
+   * @param length bytes to read
+   * @return bytes read, or -1 none of data is read
+   */
+  default int read(long position, byte[] buffer, int offset, int length)
+      throws IOException {
+    return read(position, new ByteArrayTargetBuffer(buffer, offset), length);
+  }
+
+  /**
+   * @param position position of the file to start reading data
+   * @param buffer target byte buffer
+   * @param length bytes to read
+   * @return bytes read, or -1 none of data is read
+   */
+  default int read(long position, ByteBuffer buffer, int length) throws IOException {
+    return read(position, new ByteBufferTargetBuffer(buffer), length);
+  }
+
+  /**
+   * @param position position of the file to start reading data
+   * @param buffer target byte buf
+   * @param length bytes to read
+   * @return bytes read, or -1 none of data is read
+   */
+  default int read(long position, ByteBuf buffer, int length) throws IOException {
+    return read(position, new NettyBufTargetBuffer(buffer), length);
+  }
+
+  /**
+   * @param position position of the file to start reading data
+   * @param buffer target byte buffer
+   * @param length bytes to read
+   * @return bytes read, or -1 none of data is read
+   */
+  default int read(long position, ReadTargetBuffer buffer, int length)
+      throws IOException {
+    Preconditions.checkArgument(length >= 0, "length should be non-negative");
+    Preconditions.checkArgument(position >= 0, "position should be non-negative");
+    Preconditions.checkArgument(buffer.remaining() >= length,
+        "given buffer should have enough space to write given length");
+    if (length == 0) {
+      return 0;
+    }
+    return readInternal(position, buffer, length);
+  }
+
+  /**
+   * @param position position of the file to start reading data
+   * @param buffer target byte buffer
+   * @param length bytes to read
+   * @return bytes read, or -1 none of data is read
+   */
+  int readInternal(long position, ReadTargetBuffer buffer, int length)
+      throws IOException;
```

Comment:
Do you think we should establish a contract on the state of the buffer when an exception occurs? 

For example, if the caller wants 10 MiB of data, and the first 9 MiB has been successfully written into the buffer when an exception is raised by the underlying transport. If we can get information of how many bytes have been written out of the exception, we can restart the read with the position set to the previous position + 9 MiB, so that we don't have to re-read that data.

I had a similar implementation with `PartialReadException` [here](https://github.com/dbw9580/alluxio/blob/cfdeb1b2a7ee95d316ba517e43c0cac3406b4026/core/client/fs/src/main/java/alluxio/client/file/dora/netty/PartialReadException.java). For a general interface like this, we probably cannot foresee all kinds of specific underlying causes, but we can require the implementation to provide the number of bytes read when they want to create an exception.

Reply:
Another approach would be using the `remaining` method on `ReadTargetBuffer`, and have the callers check the state of the buffer before they attempt a retry when the call throws an exception.